File: Problems/0100_0199/0122_Best_Time_to_Buy_and_Sell_Stock_II/Project_Python3/Best_Time_to_Buy_and_Sell_Stock2.py
import os
import sys
import time
from typing import List, Dict, Tuple

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # 49ms - 56ms
        profit, prev = 0, prices[0]
        for price in prices:
            if price > prev:
                profit += price - prev
            prev = price
        return profit

    def maxProfit2(self, prices: List[int]) -> int:
        # 54ms - 74ms
        profit = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                profit += prices[i] - prices[i - 1]
        return profit

    def maxProfit3(self, prices: List[int]) -> int:
        # 61ms - 62ms
        p_max, p_min = 0, sys.maxsize
        sum_max = 0
        for price in prices:
            p_min = min(p_min, price)
            profit = price - p_min
            p_max = max(p_max, profit)
            if profit > 0:
                sum_max += profit
                p_min = price
        return p_max if p_max > sum_max else sum_max

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    flds = temp.replace(", ", ",").replace("\"", "").replace("[", "").replace("]", "").rstrip()

    prices = [int(val) for val in flds.split(",")]
    print("prices = {0}".format(prices))

    sl = Solution()
    time0 = time.time()

    result = sl.maxProfit(prices)

    time1 = time.time()

    print("result = {0}".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()

File: Problems/0100_0199/0122_Best_Time_to_Buy_and_Sell_Stock_II/Project_CS/Best_Time_to_Buy_and_Sell_Stock2.cs
using System;

public class Solution
{
    public int MaxProfit(int[] prices)
    {
        // 77ms - 82ms
        int profit = 0;
        for (int i = 1; i < prices.Length; i++)
        {
            if (prices[i] > prices[i - 1])
            {
                profit += prices[i] - prices[i - 1];
            }
        }
        return profit;
    }

    public int MaxProfit2(int[] prices)
    {
        // 77ms - 86ms
        int p_max = 0;
        int p_min = int.MaxValue;
        int sum_max = 0;

        foreach (int price in prices)
        {
            p_min = Math.Min(p_min, price);
            p_max = Math.Max(p_max, price - p_min);
            if (price - p_min > 0)
            {
                sum_max += price - p_min;
                p_min = price;
            }
        }
        return p_max > sum_max ? p_max : sum_max;
    }

    public int[] str_to_int_array(string s)
    {
        if (s.Length <= 0)
            return null;

        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public void Main(string args)
    {
        string flds = args.Replace("\"", "").Replace(" ", "").Replace("[", "").Replace("]", "");
        int[] prices = str_to_int_array(flds);
        Console.WriteLine("prices[] = " + output_int_array(prices));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = MaxProfit(prices);

        sw.Stop();
        Console.WriteLine("Result = " + result.ToString());
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
